src/Databases/DatabaseOrdinary.cpp

```diff
@@ -185,6 +275,66 @@ LoadTaskPtr DatabaseOrdinary::loadTableFromMetadataAsync(
     return load_table[name.table] = makeLoadTask(async_loader, {job});
 }
 
+void DatabaseOrdinary::restoreMetadataAfterConvertingToReplicated(StoragePtr table, const QualifiedTableName & name)
+{
+    if (auto * rmt = table->as<StorageReplicatedMergeTree>())
+    {
+        auto convert_to_replicated_flag_path = getConvertToReplicatedFlagPath(getContext(), name);
+        if (!fs::exists(convert_to_replicated_flag_path))
+            return;
+
+        auto has_metadata = rmt->hasMetadataInZooKeeper();
+        if (!has_metadata.has_value())
+        {
+            LOG_INFO
+            (
+                log,
+                "No connection to ZooKeeper, can't restore metadata for {} in ZooKeeper after conversion. Run SYSTEM RESTORE REPLICA while connected to ZooKeeper.",
+                backQuote(name.getFullName())
+            );
+        }
+        else
+        {
+            if (*has_metadata)
+            {
+                LOG_INFO
+                (
+                    log,
+                    "Table {} already has metatada in ZooKeeper.",
+                    backQuote(name.getFullName())
+                );
+            }
+            else
+            {
+                try
+                {
+                    rmt->restoreMetadataInZooKeeper();
+                    LOG_INFO
+                    (
+                        log,
+                        "Metadata in ZooKeeper for {} is restored.",
+                        backQuote(name.getFullName())
+                    );
+                }
+                catch (...)
+                {
+                    /// Something unpredicted happened
+                    /// Remove flag to allow server to restart
+                    fs::remove(convert_to_replicated_flag_path);
+                    throw;
+                }
+            }
+        }
+        fs::remove(convert_to_replicated_flag_path);
```

Comment:
If we remove the flag anyway, then we can remove it before restoring metadata (just after checking if the flag exists), and remove the try-catch